Command/ReindexCommand.php

```diff
@@ -84,33 +111,124 @@ public function execute(InputInterface $input, OutputInterface $output)
                 }
             }
         }
 
-        $event = new IndexRebuildEvent($filter, $purge, $output);
-        $this->eventDispatcher->dispatch(SearchEvents::INDEX_REBUILD, $event);
+        $providerNames = $input->getOption('provider');
+
+        if (empty($providerNames)) {
+            $providers = $this->providerRegistry->getProviders();
+        } else {
+            $providers = [];
+            foreach ($providerNames as $providerName) {
+                $providers[$providerName] = $this->providerRegistry->getProvider($providerName);
+            }
+        }
+
+        foreach ($providers as $providerName => $provider) {
+            $output->writeln(sprintf('<info>provider "</>%s<info>"</>', $providerName));
+            $output->write(PHP_EOL);
+
+            foreach ($provider->getClassFqns() as $classFqn) {
+                $return = $this->reindexClass($output, $provider, $providerName, $classFqn, $batchSize);
+
+                if (false === $return) {
+                    break;
+                }
+            }
+
+            $this->resumeManager->removeCheckpoints($providerName);
+        }
 
-        $output->write(PHP_EOL);
         $output->writeln(sprintf(
-            'Index Rebuild Completed (%ss %sb)',
+            '<info>Index rebuild completed (</>%ss %sb</><info>)</>',
             number_format(microtime(true) - $startTime, 2),
             number_format(memory_get_usage())
         ));
     }
+
+    private function reindexClass(
+        OutputInterface $output,
+        ReIndexProviderInterface $provider,
+        $providerName,
+        $classFqn,
+        $batchSize
+    ) {
+        $count = $provider->getCount($classFqn);
+        $checkpoint = $this->resumeManager->getCheckpoint($providerName, $classFqn);
+        $offset = $checkpoint ?: 0;
+
+        // If the offset is the same as the count then the job was already completed.
+        // If more objects are in the database than before, then we will just index a couple
+        // of more documents than normal.
+        if ($offset === $count) {
+            return;
+        }
+
+        $realCount = $count - $offset;
+
+        $output->writeln(sprintf(
+            '<comment>-- reindexing "</>%s<comment>" instances of "</>%s<comment>"</>',
+            $realCount,
+            $classFqn
+        ));
+        $output->write(PHP_EOL);
+
+        $progress = new ProgressBar($output);
+        $progress->start($realCount);
+
+        // index in batches
+        while (true) {
+            $objects = $provider->provide($classFqn, $offset, $batchSize);
+
+            if (count($objects) === 0) {
+                break;
+            }
+
+            $this->resumeManager->setCheckpoint($providerName, $classFqn, $offset);
+
+            foreach ($objects as $object) {
+                $locales = [null];
+
+                if ($provider instanceof LocalizedReIndexProviderInterface) {
```

Comment:
The official is ReIndex. Reindex's should be removed.

On Mon, Feb 22, 2016 at 05:38:23AM -0800, Daniel Rotter wrote:
>    In [1]Command/ReindexCommand.php:
> 
>  > +        $progress->start($realCount);
>  > +
>  > +        // index in batches
>  > +        while (true) {
>  > +            $objects = $provider->provide($classFqn, $offset, $batchSize);
>  > +
>  > +            if (count($objects) === 0) {
>  > +                break;
>  > +            }
>  > +
>  > +            $this->resumeManager->setCheckpoint($providerName, $classFqn, $offset);
>  > +
>  > +            foreach ($objects as $object) {
>  > +                $locales = [null];
>  > +
>  > +                if ($provider instanceof LocalizedReIndexProviderInterface) {
> 
>    I think until now everything has been spelled as Reindex instead of
>    ReIndex.
> 
>    —
>    Reply to this email directly or [2]view it on GitHub.
> 
>    Reverse link: [3]unknown
> 
> References
> 
>    Visible links
>    1. https://github.com/massiveart/MassiveSearchBundle/pull/77#discussion_r53624881
>    2. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53624881
>    3. https://github.com/massiveart/MassiveSearchBundle/pull/77/files#r53624881